File: src/commands/here.ts
import { Command } from 'commander';
import { getContext, handleError } from './shared.js';
import { getActiveAgents, getAllAgents, getConfig, getClaimCountsByAgent } from '../db/queries.js';
import { formatRelative } from '../core/time.js';

export function hereCommand(): Command {
  return new Command('here')
    .description('List active agents')
    .option('--all', 'include stale agents')
    .action(async (options, cmd) => {
      try {
        const { db, jsonMode } = getContext(cmd);

        let agents;
        if (options.all) {
          // Show all agents that haven't explicitly left (including stale ones)
          agents = getAllAgents(db).filter(a => a.left_at === null);
        } else {
          // Show only active (non-stale, non-left) agents
          const staleHours = parseInt(getConfig(db, 'stale_hours') || '4', 10);
          agents = getActiveAgents(db, staleHours);
        }

        // Get claim counts for all agents
        const claimCounts = getClaimCountsByAgent(db);
        const messageCounts = new Map<string, number>();
        const messageCountRows = db.prepare(`
          SELECT from_agent as agent_id, COUNT(*) as count
          FROM mm_messages
          GROUP BY from_agent
        `).all() as { agent_id: string; count: number }[];
        for (const row of messageCountRows) {
          messageCounts.set(row.agent_id, row.count);
        }

        if (jsonMode) {
          const payload = {
            agents: agents.map(agent => ({
              agent_id: agent.guid,
              display_name: agent.agent_id,
              goal: agent.goal,
              last_active: new Date(agent.last_seen * 1000).toISOString(),
              message_count: messageCounts.get(agent.agent_id) || 0,
              claim_count: claimCounts.get(agent.agent_id) || 0,
            })),
            total: agents.length,
          };
          console.log(JSON.stringify(payload, null, 2));
        } else {
          if (agents.length === 0) {
            console.log('No active agents');
          } else {
            console.log(`ACTIVE AGENTS (${agents.length}):`);
            for (const agent of agents) {
              const lastSeen = formatRelative(agent.last_seen);
              const claimCount = claimCounts.get(agent.agent_id) || 0;
              const claimInfo = claimCount > 0 ? ` (${claimCount} claim${claimCount !== 1 ? 's' : ''})` : '';
              const goal = agent.goal ? ` - ${agent.goal}` : '';
              console.log(`  @${agent.agent_id}${claimInfo}${goal}`);
              console.log(`    last seen: ${lastSeen}`);
            }
          }
        }

        db.close();
      } catch (error) {
        handleError(error);
      }
    });
}

File: internal/command/here.go
package command

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/adamavenir/fray/internal/db"
	"github.com/adamavenir/fray/internal/types"
	"github.com/spf13/cobra"
)

// NewHereCmd creates the here command.
func NewHereCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "here",
		Short: "List active agents",
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			includeAll, _ := cmd.Flags().GetBool("all")

			var agents []types.Agent
			if includeAll {
				all, err := db.GetAllAgents(ctx.DB)
				if err != nil {
					return writeCommandError(cmd, err)
				}
				for _, agent := range all {
					if agent.LeftAt == nil {
						agents = append(agents, agent)
					}
				}
			} else {
				staleHours := 4
				if value, err := db.GetConfig(ctx.DB, "stale_hours"); err == nil && value != "" {
					staleHours = parseInt(value, staleHours)
				}
				agents, err = db.GetActiveAgents(ctx.DB, staleHours)
				if err != nil {
					return writeCommandError(cmd, err)
				}
			}

			claimCounts, err := db.GetClaimCountsByAgent(ctx.DB)
			if err != nil {
				return writeCommandError(cmd, err)
			}

			messageCounts, err := getMessageCounts(ctx.DB)
			if err != nil {
				return writeCommandError(cmd, err)
			}

			allRoles, err := db.GetAllAgentRoles(ctx.DB)
			if err != nil {
				return writeCommandError(cmd, err)
			}

			if ctx.JSONMode {
				payload := map[string]any{
					"agents": buildHerePayload(agents, claimCounts, messageCounts, allRoles),
					"total":  len(agents),
				}
				return json.NewEncoder(cmd.OutOrStdout()).Encode(payload)
			}

			out := cmd.OutOrStdout()
			if len(agents) == 0 {
				fmt.Fprintln(out, "No active agents")
				return nil
			}

			fmt.Fprintf(out, "ACTIVE AGENTS (%d):\n", len(agents))
			for _, agent := range agents {
				claimCount := claimCounts[agent.AgentID]
				claimInfo := ""
				if claimCount > 0 {
					plural := "s"
					if claimCount == 1 {
						plural = ""
					}
					claimInfo = fmt.Sprintf(" (%d claim%s)", claimCount, plural)
				}
				status := ""
				if agent.Status != nil && *agent.Status != "" {
					status = " - " + *agent.Status
				}
				roleInfo := ""
				if roles := allRoles[agent.AgentID]; roles != nil {
					roleInfo = formatRoleInfo(roles)
				}
				fmt.Fprintf(out, "  @%s%s%s%s\n", agent.AgentID, roleInfo, claimInfo, status)
				fmt.Fprintf(out, "    last seen: %s\n", formatRelative(agent.LastSeen))
			}

			return nil
		},
	}

	cmd.Flags().Bool("all", false, "include stale agents")
	return cmd
}

func buildHerePayload(agents []types.Agent, claimCounts map[string]int64, messageCounts map[string]int64, allRoles map[string]*types.AgentRoles) []map[string]any {
	payload := make([]map[string]any, 0, len(agents))
	for _, agent := range agents {
		entry := map[string]any{
			"agent_id":      agent.GUID,
			"display_name":  agent.AgentID,
			"status":        agent.Status,
			"last_active":   timeISO(agent.LastSeen),
			"message_count": messageCounts[agent.AgentID],
			"claim_count":   claimCounts[agent.AgentID],
		}
		if roles := allRoles[agent.AgentID]; roles != nil {
			entry["roles_held"] = roles.Held
			entry["roles_playing"] = roles.Playing
		}
		payload = append(payload, entry)
	}
	return payload
}

func formatRoleInfo(roles *types.AgentRoles) string {
	if len(roles.Held) == 0 && len(roles.Playing) == 0 {
		return ""
	}
	allRoles := append([]string{}, roles.Held...)
	for _, r := range roles.Playing {
		allRoles = append(allRoles, r+"*")
	}
	return " [" + strings.Join(allRoles, ", ") + "]"
}

func timeISO(ts int64) string {
	return time.Unix(ts, 0).UTC().Format(time.RFC3339)
}
